types/eventbus/event_bus.go

```diff
@@ -0,0 +1,241 @@
+package eventbus
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
+	"github.com/tendermint/tendermint/libs/service"
+	"github.com/tendermint/tendermint/types"
+)
+
+type Subscription interface {
+	ID() string
+	Next(context.Context) (tmpubsub.Message, error)
+}
+
+// EventBus is a common bus for all events going through the system. All calls
+// are proxied to underlying pubsub server. All events must be published using
+// EventBus to ensure correct data types.
+type EventBus struct {
+	service.BaseService
+	pubsub *tmpubsub.Server
+}
+
+// NewDefault returns a new event bus with default options.
+func NewDefault() *EventBus {
+	pubsub := tmpubsub.NewServer(tmpubsub.BufferCapacity(0))
+	b := &EventBus{pubsub: pubsub}
+	b.BaseService = *service.NewBaseService(nil, "EventBus", b)
+	return b
+}
+
+func (b *EventBus) SetLogger(l log.Logger) {
+	b.BaseService.SetLogger(l)
+	b.pubsub.SetLogger(l.With("module", "pubsub"))
+}
```

Comment:
I bet we can get rid of the set logger if we make sure to pass them in during initializing.

Reply:
The pubsub service doesn't have a constructor option for setting the logger. I could add one, but I kinda didn't want to mess around with the API there yet. I would like to go back in there and get rid of the functional options pattern and clean that up—though I think maybe that were better done separately?